### PR TITLE
Precision strike

### PR DESCRIPTION
**Pull request by lheckemann**
###### Motivation for this change
#21852 It would be nice to allow removing specific references rather than all references besides the ones specifically listed.

Also includes a sample usage for syncthing. I think this should be used in more places (e.g. buildGoPackage, perhaps some places that currently use nuke-refs), but was hoping to get some feedback on the idea, the implementation and the name (chosen by analogy to “nuke”).

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/precision-strike/builder.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+targets=()
+while getopts t: o; do
+    case "\$o" in
+        t) storeId=\$(echo "\$OPTARG" | sed -n "s|^$NIX_STORE/\\([a-z0-9]\{32\}\\)-.*|\1|p")
+           if [ -z "\$storeId" ]; then
+               echo "-t argument must be a Nix store path"
+               exit 1
+           fi
+           targets+=("$storeId")
```
Comment (joachifm) [inline pkgs/build-support/precision-strike/builder.sh]: Here `$storeId` is not escaped, is that intentional?

Comment (lheckemann) [inline pkgs/build-support/precision-strike/builder.sh]: Good call, fixed.

**Comment by abbradar** [comment]
No problems at first glance, thanks! I like the name :D

I'll merge in several days if noone beats me to it.

**Comment by lheckemann** [comment]
Turns out it's actually rather broken because apparently I don't test this kind of stuff properly. Hold on the merge...

**Comment by lheckemann** [comment]
Fixed, now happy to have it merged.

**Comment by joachifm** [comment]
I'd like to merge but can't help but ask if we could do away with the perl dep and just use sed for the substitution, or is there some special sauce in the perl invocation not easily replicable with sed?

**Comment by lheckemann** [comment]
@joachifm done. This assumes it's GNU sed, I hope that's okay. This is because of inconsistency across sed implementations in terms of handling the `-i` flag — I remember encountering issues with using `-i` on OSX sed in the past.

**Comment by joachifm** [comment]
From what I can tell, all the stdenvs use the GNU tools so it should be okay (`sed -i` is used a lot already). It's an unfortunate inconsistency though ...

**Comment by pshendry** [comment]
Just my two cents: I don't think tools like these should have novelty names. Something like `pruneReferences` would more accurately describe what it does, particularly in contrast to the similar `nukeReferences`.

**Comment by lheckemann** [comment]
I do agree that the novelty name feels inappropriate. However, I think in this case it describes the functionality a bit better than something like `pruneReferences` which to me doesn't convey how it's targetting very specific references very particularly. Definitely open to a better name though.

**Comment by bjornfor** [comment]
killReferences?

**Comment by pshendry** [comment]
`excludeReferences`?

**Comment by lheckemann** [comment]
>killReferences?

Also doesn't really convey that it's killing specific references rather than all of them.

[Precision Ref Strike](http://tvtropes.org/pmwiki/pmwiki.php/Main/PrecisionFStrike). [sorry…]

**Comment by bjornfor** [comment]
removeReferences?

**Comment by joachifm** [comment]
`removeReferencesTo`

**Comment by lheckemann** [comment]
Changed name to removeReferencesTo, rebased on master, and switched to using `writeScriptBin` (a change that might make sense for nuke-refs as well).

**Comment by joachifm** [comment]
This LGTM. An idea for the future: what if there was a `stdenv` attribute called `removeReferencesTo` and a standard setup hook that'd process all outputs and strip out those references automatically, obviating the need to pass `removeReferencesTo` explicitly & add that boiler plate to `postFixup`?

**Comment by lheckemann** [comment]
Yes, I like that! That also allows expressing the outputs as a nix list rather than manual calls to `removeReferencesTo`. The more declarative, the better!

**Comment by joachifm** [comment]
Seems to me to work as advertised. Thank you

**Comment by joachifm** [comment]
@lheckemann this is an awful hackup of what it could look like (sans stdenv integration, so the hook still needs to be added to inputs manually): https://github.com/joachifm/nixpkgs/tree/remove-reference-to-setup-hook An actual implementation I think would end up unifying the two and add some checks for robustness and whatnot, but the demo seems to work.

Not sure if it'll be used enough to warrant this sort of integration, though it'd save quite a bit of boilerplate.

**Comment by Profpatsch** [comment]
This copies the builder of `nuke-references` nearly verbatim. Is there no way to factor that out? @lheckemann

**Comment by lheckemann** [comment]
The only code that's almost copied verbatim is the option-getting part, and the code implementing the actual functionality is very different. I don't think it would be very beneficial to make the option-getting code common to the two.

**Pull request merged**